File: core/src/io/github/jerukan/util/shapes/Shape.java
package io.github.jerukan.util.shapes;

import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.math.Vector2;

/**
 * Generic shape that holds only its position.
 */

public abstract class Shape {

    public Vector2 position;

    public Shape(Vector2 pos) {
        position = pos;
    }

    public abstract boolean collidesPoint(Vector2 other);

    public abstract boolean collidesRect(Rectangle other);

    public abstract boolean collidesCircle(Circle other);

    public boolean collides(Shape other) {
        if(other instanceof Circle) {
            return collidesCircle((Circle)other);
        }
        else if(other instanceof Rectangle) {
            return collidesRect((Rectangle) other);
        }
        return false;
    }

    public void setPosition(float x, float y) {
        position.set(x, y);
    }

    public void translate(float x, float y) {
        position.add(x, y);
    }

    public abstract void scale(float magnitude);

    public abstract float getArea();

    public abstract float getPerimeter();

    public abstract void render(ShapeRenderer shapeRenderer);
}

File: core/src/io/github/jerukan/util/shapes/Shape.kt
package io.github.jerukan.util.shapes

import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import com.badlogic.gdx.math.Vector2

/**
 * Generic shape that holds only its position.
 */

abstract class Shape(pos: Vector2) {

    var position: Vector2 = pos
    set(value) {
        setPosition(value.x, value.y)
    }

    open fun setPosition(x: Float, y: Float) {
        position.set(x, y)
    }

    abstract fun collidesPoint(other: Vector2): Boolean

    abstract fun collidesRect(other: Rectangle): Boolean

    abstract fun collidesCircle(other: Circle): Boolean

    fun collides(other: Shape): Boolean {
        if(other is Circle) {
            return collidesCircle(other)
        }
        else if(other is Rectangle) {
            return collidesRect(other)
        }
        return false
    }

    open fun translate(x: Float, y: Float) {
        position.add(x, y)
    }

    abstract fun scale(magnitude: Float)

    abstract fun getArea(): Float

    abstract fun getPerimeter(): Float

    abstract fun render(shapeRenderer: ShapeRenderer)
}